### PR TITLE
client, lang, common, templates: Replace illyadance with customizable video background

### LINKED ISSUE #963: happi flens
Should happi flens be a client-side config option like illya dance?
Everyone seems to love it.

### PR DESCRIPTION
**Pull request by Chiiruno**
Fixes #963
Seems to work, only problem I'm having is that the lang text isn't showing up properly for the "new" options. Any ideas why?

### CODE REVIEW

**Comment by Chiiruno** [comment]
Of course, illya dance should stay, just move her to the videos folder, it will auto detect.

**Comment by Chiiruno** [comment]
Ah, but you will have to restart the server every time you add a video, at least until I figure out a sane way to detect over time, maybe a loop with a timer, dunno.
Having it detect on folder change would be nice.

**Comment by Robbotto** [comment]
Hi,

Please unsubscribe me to this thread.  I have already ask before and I am still receiving messages.

I look forward to hearing from you, soon.

thanks.

Jaime

**Comment by Robbotto** [comment]
Please unsubscribe me.

Thanks,

Jaime

**Comment by Chiiruno** [comment]
Thanks for the tips, lat.
Is this guy a troll? This is the second time this has happened.

**Comment by Chiiruno** [comment]
At any rate, "Jaime", you need to unfollow @bakape if you want to stop these notifications, or, turn off email notifications if they are bothering you.
Please read up on how GitHub works before spamming threads, thank you.

**Comment by bakape** [comment]
@Chiiruno it's a bot. Report, block and forget.

**Comment by Chiiruno** [comment]
Oh, I feel dumb now.
I'll hide all related posts I guess.

**Comment by Chiiruno** [comment]
Okay, I think I might have a good flow for this.
Watch www/videos directory on server init with fsnotify -> rebuild templates on change (not sure how to rebuild) -> move `FetchVideoNames` to cache instead of common, protect with mutex.
Does that sound about right?

**Comment by bakape** [comment]
>not sure how to rebuild
See `db/config.go`. You should probably move the function there to the
templates module.
>move FetchVideoNames to cache
No, just have a private global variable there protected by a RW mutex.

**Comment by bakape** [comment]
That said,  `FetchVideoNames()` should probably be moved from `/common` to
`/assets`.

**Comment by bakape** [comment]
Does restarting the server propagate the updates to the client?

**Comment by Chiiruno** [comment]
Yes.

**Comment by Chiiruno** [comment]
But, only if the client refreshes.
It's not completely live.

**Comment by bakape** [comment]
Then either:
a) You aren't actually recompiling the template.
b) Fuck, the page etag does not change, because  it only takes configs into account. 
  b.1) Append a hash of this to the etag of all pages? I can't think of a better solution right now.

**Comment by bakape** [comment]
Also, you need to purge the entire cache  - not good.

**Comment by bakape** [comment]
>  if you want to take some time to figure out a non-shitty and performant
way of fixing what seems to be an overlaying issue
Store hash on generation and append it to all HTML etags.

**Comment by Chiiruno** [comment]
This is a new area for me, if you think that's the best way, you'll have to point me in the right direction.

**Comment by Chiiruno** [comment]
Okay, I wasn't changing the optionSpec variable at the right time, all I need in that respect now is to figure out a better way to change the variable.
Other than that, all that's left is why the lang text isn't correct client-wise.
screenshot_20190121_005507

**Comment by bakape** [comment]
Don't rush. I won't be able to review it until tomorrow evening.

**Comment by bakape** [comment]
@Chiiruno Tell me, when it's ready.

**Comment by Chiiruno** [comment]
@bakape I've been waiting for you to help me with the two questions I asked.
I'm at a bit of an impasse.

**Comment by bakape** [comment]
Which ones?

**Comment by Chiiruno** [comment]
>Okay, I wasn't changing the optionSpec variable at the right time, all I need in that respect now is to figure out a better way to change the variable.
>Other than that, all that's left is why the lang text isn't correct client-wise.

Please see the image a few posts above also.

**Comment by bakape** [comment]
Fuck

**Comment by bakape** [comment]
Okay, the first template compilation happens only on server start. You will need to inject a new variable into the template as seen in https://github.com/bakape/meguca/blob/master/server/src/meguca/templates/index.qtpl#L42-L46 .
Then you will need to change the select element in the option manually after client init.

**Comment by bakape** [comment]
>why the lang text isn't correct client-wise

Run `make generate`.

**Comment by Chiiruno** [comment]
>Run `make generate`.

Oh, it doesn't do that automatically anymore in the makefile.
Thanks, I would have really never, ever thought to look there.

**Comment by bakape** [comment]
>Oh, it doesn't do that automatically anymore in the makefile.

Preparation for Go modules.

**Comment by Chiiruno** [comment]
Okay, I'm not having a problem with functionality, I already have that taken care of.
The thing I don't like is changing the optionSpec by array indexing.
See: https://github.com/bakape/meguca/blob/d1f426a52776f1775d366da75e99c051d68ea744/server/src/meguca/templates/templates.go#L31-L34

**Comment by bakape** [comment]
You no longer need to touch the spec. I forgot that it's supposed to be
immutable. Patch on the client side only.

**Comment by Chiiruno** [comment]
Okay, so I need to send an extra piece of information, where you defined a few posts above, then?

**Comment by bakape** [comment]
Yes. JSON array is probably best. See how `boards` is done.

**Comment by Chiiruno** [comment]
@bakape Okay, I have it completely working again now.
However, there is a duplication I'm not sure how to fix, but I believe it is because of where I placed the generation logic.
Basically, this happens.
screenshot_20190123_050351

**Comment by Chiiruno** [comment]
@bakape Ready for merge, pending your review.

File: client/options/background.ts [hunk 1]
```diff
@@ -48,8 +48,8 @@ export function render(bg?: BackgroundStore) {
 	style.innerHTML = ''
 
 	let showOPBG = false
-	if (options.illyaDance) {
-		renderIllya()
+	if (options.bgVideo != 'none') {
```
Comment (bakape) [inline client/options/background.ts]: `!==` :^)

File: client/options/background.ts [hunk 2]
```diff
@@ -58,18 +58,12 @@ export function render(bg?: BackgroundStore) {
...
-	}
-	const html = HTML
-		`<video ${args}>
-			<source src="/assets/illya.webm" type="video/webm">
-			<source src="/assets/illya.mp4" type="video/mp4">
+// Attach video to the background
+function renderBgVideo() {
+	container.innerHTML = HTML
+		`<video autoplay loop${options.bgMute ? ' muted' : ''}>
+			<source src="/assets/videos/${options.bgVideo}" type="${options.bgVideo.split('.').pop() == "webm" ? "video/webm" : "video/mp4"}">
```
Comment (bakape) [inline client/options/background.ts]: `===` :^)

File: client/options/index.ts [hunk 3]
```diff
@@ -146,9 +147,23 @@ export function initOptions() {
 		new OptionModel(id, specs[id])
 	}
 
+	// Manually change bgVideo select as it is dynamically generated
+	const bgElement = document.getElementById("bgVideo")
+
+	for (let val of bgElement.getElementsByTagName("option")) {
```
Comment (bakape) [inline client/options/index.ts]: `bgElement.innerHTML = ""` is cleaner.

File: client/options/specs.ts [hunk 4]
```diff
@@ -103,20 +103,22 @@ export const specs: { [id: string]: OptionSpec } = {
...
 		noExecOnStart: true,
 		exec: renderBackground,
 	},
-	// Mute Illya dance
-	illyaDanceMute: {
+	// Random video player
+	meguTV: {
 		noExecOnStart: true,
-		exec: renderBackground,
+		exec: initTV,
```
Comment (bakape) [inline client/options/specs.ts]: Why the change?

File: server/src/meguca/assets/common.go [hunk 5]
```diff
@@ -1,8 +1,29 @@
...
+	"github.com/fsnotify/fsnotify"
+	"github.com/go-playground/log"
+
+	"meguca/common"
+	"meguca/util"
+)
+
+// Export to avoid circular dependency
+func init() {
+	common.GetVideoNames = GetVideoNames
```
Comment (bakape) [inline server/src/meguca/assets/common.go]: Why do we need this instead of just importing `assets` from `templates`?

File: server/src/meguca/assets/common.go [hunk 6]
```diff
@@ -45,3 +66,83 @@ func (s *FileStore) Get(board string) (file File) {
...
 	return file
 }
+
+func generateVideoNames() {
+	rw.Lock()
+	defer rw.Unlock()
+
+	ret := []string{"none"}
+	files, err := ioutil.ReadDir("www/videos")
+
```
Comment (bakape) [inline server/src/meguca/assets/common.go]: You have some odd fetishes.

File: server/src/meguca/assets/common.go [hunk 7]
```diff
@@ -45,3 +66,83 @@ func (s *FileStore) Get(board string) (file File) {
 	}
 	return file
 }
+
+func generateVideoNames() {
```
Comment (bakape) [inline server/src/meguca/assets/common.go]: Needs error handling.

File: server/src/meguca/server/daemon.go [hunk 8]
```diff
@@ -26,6 +28,7 @@ func init() {
 		case "debug":
 			mlog.Init(mlog.Console)
 			mlog.ConsoleHandler.SetDisplayColor(true)
+			go assets.WatchVideoDir()
```
Comment (bakape) [inline server/src/meguca/server/daemon.go]: Put this into `startSever()` with the other initializers.

File: server/src/meguca/server/daemon.go [hunk 9]
```diff
@@ -38,6 +41,7 @@ func init() {
 		case "start":
 			mlog.Init(mlog.Console)
 			mlog.ConsoleHandler.SetDisplayColor(false)
+			go assets.WatchVideoDir()
```
Comment (bakape) [inline server/src/meguca/server/daemon.go]: Delete this.

File: client/options/specs.ts (hunk 4 above)
Comment (Chiiruno) [reply client/options/specs.ts]: That's not actually changed, it's github being weird with showing the revision changes.
Please see the actual file.

File: server/src/meguca/assets/common.go (hunk 6 above)
Comment (Chiiruno) [reply server/src/meguca/assets/common.go]: What?

File: client/options/index.ts (hunk 3 above)
Comment (Chiiruno) [reply client/options/index.ts]: I think I remember doing that before, but decided not to for some reason.
I'll try again, maybe there was another issue.

File: server/src/meguca/assets/common.go (hunk 5 above)
Comment (Chiiruno) [reply server/src/meguca/assets/common.go]: Circular dependencies, need to be able to recompile templates from assets module.

Comment (bakape) [reply server/src/meguca/assets/common.go]: Oh, Okay.

File: server/src/meguca/assets/common.go (hunk 7 above)
Comment (bakape) [reply server/src/meguca/assets/common.go]: Oh, there is.

File: server/src/meguca/assets/common.go (hunk 6 above)
Comment (Chiiruno) [reply server/src/meguca/assets/common.go]: Oh, good point, no need because it's already locked by a mutex, I can just use the `videoNames` itself.

File: server/src/meguca/assets/common.go (hunk 7 above)
Comment (Chiiruno) [reply server/src/meguca/assets/common.go]: Okay, we can avoid recompilation if `err != nil`.

Comment (Chiiruno) [reply server/src/meguca/assets/common.go]: Wait, no, I want to keep recompilation, because if the directory is empty or doesn't exist, the client should show [none].

**Comment by bakape** [comment]
OOOOOOOOOOOOOOOOkay, I can't compile the client.

**Comment by bakape** [comment]
@Chiiruno Also, some lazy fag pointed https://openings.moe/hub/dev/api.php out.

**Comment by Chiiruno** [comment]
Ah yes, I didn't know about that.
That said, the less meguca depends on other services, the better.

**Comment by bakape** [comment]
@Chiiruno Good point.
